controllers/limitador_status.go

```diff
@@ -115,25 +116,35 @@ func (r *LimitadorReconciler) checkLimitadorAvailable(ctx context.Context, limit
 		Name:      limitador.DeploymentName(limitadorObj),
 	}
 	err := r.Client().Get(ctx, dKey, deployment)
-	if err != nil && !apierrors.IsNotFound(err) {
+	if client.IgnoreNotFound(err) != nil {
 		return nil, err
 	}
 
-	if err != nil && apierrors.IsNotFound(err) {
-		tmp := err.Error()
-		return &tmp, nil
+	if apierrors.IsNotFound(err) {
+		return ptr.To(err.Error()), nil
 	}
 
-	availableCondition := helpers.FindDeploymentStatusCondition(deployment.Status.Conditions, "Available")
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		return ptr.To("Deployment still in progress"), nil
+	}
+
+	availableCondition := helpers.FindDeploymentStatusCondition(deployment.Status.Conditions, string(appsv1.DeploymentAvailable))
 	if availableCondition == nil {
-		tmp := "Available condition not found"
-		return &tmp, nil
+		return ptr.To("Available condition not found"), nil
 	}
 
 	if availableCondition.Status != corev1.ConditionTrue {
 		return &availableCondition.Message, nil
 	}
 
+	if deployment.Status.UnavailableReplicas != 0 {
+		return ptr.To("Deployment has unavailable replicas"), nil
+	}
+
+	if deployment.Status.ReadyReplicas != deployment.Status.Replicas {
+		return ptr.To("Deployment has replicas not ready yet"), nil
+	}
+
```

Comment:
Any thoughts about an edge case where an user sets `limitador.spec.replicas` to `0`, so `deployment.status.replicas` is `0`. The current logic will mark the CR status as `Ready: true`, would this be the expected status for this edge case? 🤔

Reply:
good point. I think we should follow the same approach as the replication controller for the Deployment object.

I set limtiador CR with replica: 0

```yaml
spec:
  replicas: 0
```

The operator updates the deployment object to:
```yaml
spec:
  replicas: 0
  ....
```

And the status of the deployment is:

```yaml
status:
  conditions:
  - lastTransitionTime: "2024-06-10T08:26:11Z"
    lastUpdateTime: "2024-06-10T08:26:11Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2024-06-10T08:25:51Z"
    lastUpdateTime: "2024-06-10T08:26:11Z"
    message: ReplicaSet "limitador-limitador-76c6d7bd46" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 2
```

Note that as `status.availableReplicas`, `status.readyReplicas`, `status.replicas` are all `0`, they are not shown in the response (considered "empty").

Yet, the status condition `Available` is `True`. Thus, we can account that as "available". 

Thus, IMO, limitador CR status should also be considered "Ready". I agree, though, that "Ready" does not reflect the actual state. It would be more meaningful, "Reconciled" or "Stable" or something that means, limitador is what it should be, what the user wanted to be. When the replicas are `0`, I can leave "Ready" to "True", and the message would be "Limitador is ready, but without running instances". (or something like that).